cg/smart_contracts/ClearGDPR.sol

```diff
@@ -1,135 +1,152 @@
+        for(uint256 i = 0; i < _processors.length; i++){
+            require(isProcessor(_processors[i]));
+        }
+        return true;
     }
-    return true;
-  }
   
-  function recordProcessorsUpdate(address[] _newProcessors) public onlyController returns(bool){
-    require(setProcessors(_newProcessors));
-    emit Controller_ProcessorsUpdated(_newProcessors);  
-    return true;
-  }
+    function recordProcessorsUpdate(address[] _newProcessors) public onlyController returns(bool){
+        require(setProcessors(_newProcessors));
+        emit Controller_ProcessorsUpdated(_newProcessors);  
+        return true;
+    }
 
-  function recordConsentGivenTo(bytes32 _subjectIdHash, address[] _processorsWhiteListedBySubject) public onlyController notErased(_subjectIdHash)  returns (bool){
-    require(areAllValidProcessors(_processorsWhiteListedBySubject));
-    for(uint256 i = 0; i < processors.length; i++){
-      subjects[_subjectIdHash].statePerProcessor[processors[i]] = State.notShareable;  
-    }
-    for(i = 0; i < _processorsWhiteListedBySubject.length; i++){
-      subjects[_subjectIdHash].statePerProcessor[_processorsWhiteListedBySubject[i]] = State.shareable;  
-    }
-    subjects[_subjectIdHash].isErased = false;
-    emit Controller_ConsentGivenTo(_subjectIdHash, _processorsWhiteListedBySubject);
-    return true;
-  }
+    function recordConsentGivenTo(bytes32 _subjectId, address[] _processorsConsented) public onlyController notErased(_subjectId) returns(bool){
+        require(areAllValidProcessors(_processorsConsented));
+        for(uint256 i = 0; i < processors.length; i++){
+            subjects[_subjectId].statePerProcessor[processors[i]] = State.notShareable;  
+        }
+        for(i = 0; i < _processorsConsented.length; i++){
+            subjects[_subjectId].statePerProcessor[_processorsConsented[i]] = State.shareable;  
+        }
+        subjects[_subjectId].statePerProcessor[controller] = State.shareable;  
+        subjects[_subjectId].isErased = false;
+        emit Controller_ConsentGivenTo(_subjectId, _processorsConsented);
+        return true;
+    }
   
-  function recordAccessByController(bytes32 _subjectIdHash) public onlyController notErased(_subjectIdHash) returns (State[]){
-    State[] memory states = new State[](processors.length);
-    for(uint256 i = 0; i < processors.length; i++){
-      states[i] = subjects[_subjectIdHash].statePerProcessor[processors[i]];
-    }
-    emit Controller_SubjectDataAccessed(_subjectIdHash);
-    return states;
-  }
+    function recordAccessByController(bytes32 _subjectId) public onlyController notErased(_subjectId) returns(State[]){
+        State[] memory states = new State[](processors.length);
+        for(uint256 i = 0; i < processors.length; i++){
+            states[i] = subjects[_subjectId].statePerProcessor[processors[i]];
+        }
+        emit Controller_SubjectDataAccessed(_subjectId);
+        return states;
```

Comment:
Not sure what we need these states for. Won't this just return an array like:
`[ shareable, notsherable, erased, erase, ... ]`
???

Reply:
Yeah, we can work some more in this. The idea is to get the state of the subject's data for each processor it has given consent. A better response would be:

{
  processor1: notShareable
  processor2: erased
  ...
}

But since it's difficult to return JSON in Solidity, although I think it is possible, were returning only the states for each processor. A simple change would be return the array with the current processors in the system along with the array with its respective state. then the back-end could iterate over the arrays, creating a nice JSON for the front-end